File: leetcode/230.kth-smallest-element-in-a-bst.py
# @leet start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: TreeNode | None, k: int) -> int:
        res = -1

        def inorder(root: TreeNode | None) -> bool:
            nonlocal k, res

            if not root:
                return False

            if inorder(root.left):
                return True

            k -= 1

            if k == 0:
                res = root.val

                return True

            return inorder(root.right)

        inorder(root)

        return res


# @leet end

File: leetcode/230.kth-smallest-element-in-a-bst.js
// @leet start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function (root, k) {
  let currCount = 0;

  const inOrder = (node) => {
    if (node === null) {
      return null;
    }

    const left = inOrder(node.left);

    if (left !== null) {
      return left;
    }

    currCount++;

    if (currCount === k) {
      return node.val;
    }

    return inOrder(node.right);
  };

  return inOrder(root);
};
// @leet end
